PRESIDECMS-2213 batch edit delete background task

DominicWatson opened the pull request:
Puts batch edit/delete operations into an adhoc task-based background thread. 

Note: further refactoring will be done against PRESIDECMS-2214 where we introduce filter based batch edit/delete.

DominicWatson committed:
[twgit] Init feature 'feature-PRESIDECMS-2213_batch-edit-delete-background-task'.

DominicWatson committed:
PRESIDECMS-2213 use a bg task with progress monitor for doing datamanager batch record updates

DominicWatson committed:
PRESIDECMS-2213 remove transaction around batch update

It really is not strictly necessary, will cause big problems with large updates + stops the log from updating while progress is made

Also fix bad variable name that had gone unnoticed for a long time (worked previously because the variable was available in URL scope)

DominicWatson committed:
PRESIDECMS-2213 refactor delete record method in datamanager handler

Use separate handler for bulk delete and tidy up the single record delete (still to do: implement the separate batch delete service method)

DominicWatson committed:
PRESIDECMS-2213 implement batch delete in background thread + tweaks to batch update to make auditing work

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/pixl8/Preside-CMS?pullRequest=1036) <br/>All committers have signed the CLA.

DominicWatson committed:
[twgit] Init feature 'feature-PRESIDECMS-2214_batch-all-matching-current-filter'.

DominicWatson committed:
Merge branch 'feature-PRESIDECMS-2213_batch-edit-delete-background-task' into feature-PRESIDECMS-2214_batch-all-matching-current-filter

DominicWatson committed:
PRESIDECMS-2214 HTML and javascript behaviour for showing option to select all filtered records when doing batch operations

Still to do: actually submit the filter + use that in batch operations

DominicWatson committed:
PRESIDECMS-2214 frontend logic to prepare batch action form with filter data when selecting all records

Stil TODO: actual server side logic to make this work :) :) :)

DominicWatson committed:
PRESIDECMS-2214 return args used to fetch data when fetching data for grid tables

DominicWatson committed:
PRESIDECMS-2214 store hash of source string to prevent tampering with it later (we will use this to validate the input when using for batch operations)

DominicWatson committed:
PRESIDECMS-2214 rename variable inline with naming convention of other returned data

DominicWatson committed:
PRESIDECMS-2214 refactor frontend implementation to use returned data source string for batch operation submission

DominicWatson committed:
PRESIDECMS-2214 validate batch source in multi action handler

DominicWatson committed:
PRESIDECMS-2214 allow adhoc tasks to be interruptable with the existing toothless cancel task function

DominicWatson committed:
2214 implement batch all delete

DominicWatson committed:
PRESIDECMS-2214 simplify logging

and fix little logical bug with do/while loop clause

DominicWatson committed:
PRESIDECMS-2214 Implement batch edit using 'select all records' feature

DominicWatson committed:
PRESIDECMS-2214 rename batch_edit_queue to batch_operation_queue so as to be more useful

DominicWatson committed:
PRESIDECMS-2214 provide an API into operation queue

DominicWatson committed:
PRESIDECMS-2214 add bg task to cleanup old queued batch operations

DominicWatson committed:
PRESIDECMS-2214 fix missing elvis for uncertain url param

DominicWatson committed:
PRESIDECMS-2214 separate out service layer for batch operations into its own service

DominicWatson committed:
PRESIDECMS-2214 documentation tweaks

DominicWatson committed:
PRESIDECMS-2214 make batch delete use the operation queue rather than evaluating filter on each loop (avoids potential for endless loop)

DominicWatson committed:
PRESIDECMS-2214 normalize variable names

DominicWatson committed:
PRESIDECMS-2214 update i18n to indicate that batch operation will apply to roughly the number of records shown

DominicWatson committed:
PRESIDECMS-2214 remove unwanted variable

DominicWatson committed:
PRESIDECMS-2214 tweaks to make batch delete operation have its own pre/post hooks that code should update to (doesn't make sense to use the old hooks)

Also optimize the queue fetching so that we delete items from the queue as soon as we fetch them (by default)

DominicWatson committed:
PRESIDECMS-2214 fix and add test

DominicWatson committed:
PRESIDECMS-2214 just use default pagination argument in our calls

DominicWatson committed:
PRESIDECMS-2214 show numbers of cancelled queued operations when cancelling out of a task early

DominicWatson committed:
PRESIDECMS-2214 show actual processed count for final log message on record updates

DominicWatson committed:
PRESIDECMS-2214 ensure that processed variable is incremented no matter what scenario we're in

DominicWatson committed:
Merge branch 'feature-PRESIDECMS-2214_batch-all-matching-current-filter' into feature-PRESIDECMS-2213_batch-edit-delete-background-task

DominicWatson committed:
PRESIDECMS-2213 add some backward compatibility variables for old code using pre/post delete customisations
